In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('dataset/urbania_dataset.csv')
df.head()

id                                              title   price_S  price_USD  \
0  1  Proyecto inmobilario en venta en Calle Coronel...  S/445388  USD120375   
1  2  Proyecto inmobilario en venta en Calle Coronel...  S/441323  USD119277   
2  3  Proyecto inmobilario en venta en Calle Coronel...  S/600499  USD162297   
3  4  Proyecto inmobilario en venta en Calle Coronel...  S/583916  USD157815   
4  5  Proyecto inmobilario en venta en Jirón Batalló...  S/887994        NaN   

                                             address  property_type  \
0  Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    
1  Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    
2  Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    
3  Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    
4  Jirón Batallón Callao Norte  , Santa TeresaSan...  Departamento    

                                          image_urls  \
0  {https://img10.naventcdn.com/avisos/resize/111...   
1  {https://img10.naventcdn.com/avisos/resize/111...   
2  {https://img10.naventcdn.com/avisos/resize/111...   
3  {https://img10.naventcdn.com/avisos/resize/111...   
4  {https://img10.naventcdn.com/avisos/resize/111...   

                                         description area_total area_roof  \
0  Mozz se deriva de la palabra en italiano mozio...        NaN      45m²   
1  Mozz se deriva de la palabra en italiano mozio...       45m²      45m²   
2  Mozz se deriva de la palabra en italiano mozio...       60m²      60m²   
3  Mozz se deriva de la palabra en italiano mozio...       60m²      60m²   
4  Vive exclusivo en Surco | Vive en Libitum | En...      110m²     110m²   

  bath  bed  parking   antiquity  \
0  NaN  NaN      NaN  A estrenar   
1  NaN  NaN      NaN  A estrenar   
2    2  2.0      NaN  A estrenar   
3    2  2.0      NaN  A estrenar   
4    3  3.0      NaN  A estrenar   

                                    general_features          services  \
0  {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
1  {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
2  {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
3  {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
4  {"No. de edificios que conforman el proyecto",...               NaN   

  ambience                                       property_url  
0      NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
1      NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
2      NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
3      NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
4      NaN  https://urbania.pe/inmueble/proyecto/ememvein-...

In [3]:
df['id'] = [x for x in range(df['id'].shape[0])]
df.set_index('id', inplace=True)
df.head(2)

title   price_S  price_USD  \
id                                                                           
0   Proyecto inmobilario en venta en Calle Coronel...  S/445388  USD120375   
1   Proyecto inmobilario en venta en Calle Coronel...  S/441323  USD119277   

                                              address  property_type  \
id                                                                     
0   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    
1   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento    

                                           image_urls  \
id                                                      
0   {https://img10.naventcdn.com/avisos/resize/111...   
1   {https://img10.naventcdn.com/avisos/resize/111...   

                                          description area_total area_roof  \
id                                                                           
0   Mozz se deriva de la palabra en italiano mozio...        NaN      45m²   
1   Mozz se deriva de la palabra en italiano mozio...       45m²      45m²   

   bath  bed  parking   antiquity  \
id                                  
0   NaN  NaN      NaN  A estrenar   
1   NaN  NaN      NaN  A estrenar   

                                     general_features          services  \
id                                                                        
0   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
1   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   

   ambience                                       property_url  
id                                                              
0       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
1       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...

In [4]:
df.tail(2)

title   price_S  price_USD  \
id                                                                         
22414                 Venta Dpto 1er Piso y Cochera  S/383250  USD105000   
22415  Moderno Flat de 2 Dorm. en Magdalena del Mar  S/703000  USD190000   

                                       address  property_type  \
id                                                              
22414  Calle Monsefú 189, San MiguelLima, Lima  Departamento    
22415      Juan de Aliaga, MagdalenaLima, Lima  Departamento    

                                              image_urls  \
id                                                         
22414  {https://img10.naventcdn.com/avisos/resize/111...   
22415  {https://img10.naventcdn.com/avisos/resize/111...   

                                             description area_total area_roof  \
id                                                                              
22414  ÁREA: 70 M2 <br><br>Ubicado en calle Monsefú c...       70m²      70m²   
22415  **Ideal para inversión** <br><br>Precioso flat...       87m²      90m²   

      bath  bed  parking antiquity  \
id                                   
22414    2  2.0      NaN         8   
22415    2  2.0      NaN         4   

                                        general_features  \
id                                                         
22414  {"Sistema contra incendios","Reposteros en coc...   
22415  {"Sistema contra incendios","Reposteros en coc...   

                                                services  \
id                                                         
22414  {"Video vigilancia","Area de lavandería","Sist...   
22415  {"Video vigilancia","Areas verdes","Línea blan...   

                                                ambience  \
id                                                         
22414          {Desagüe,"Dormitorio principal con baño"}   
22415  {Desagüe,"Sala de estar","Hall de ingreso","Ar...   

                                            property_url  
id                                                        
22414  https://urbania.pe:443/inmueble/clasificado/ve...  
22415  https://urbania.pe:443/inmueble/clasificado/ve...

We are interested into housing sector, i.e., the property type like 'Departamento', 'Casa', Casa de Playa', Casa de campo', 'Habitacion'.

In [5]:
df['property_type'] = df['property_type'].apply(lambda x: str(x).strip())
df['property_type'].unique()

array(['Departamento', 'Casa', 'Terreno / Lote', 'Oficina',
       'Local industrial', 'Local comercial', 'Hotel', 'Casa de playa',
       'Casa de campo', 'Habitación', 'Otros', 'Garage',
       'Terreno agrícola', 'nan'], dtype=object)

As we see we need to ignore that values of 'Terreno / Lote', 'Oficina', 'Local inductrial', 'Local comercial', 'Hotel', 'Garage', 'Terreno agrícola'.

In [6]:
housing_values = ['Departamento', 'Casa', 'Casa de playa', 'Casa de campo',
                    'Habitación', 'Otros']
mask = df['property_type'].apply(lambda x: x in housing_values)
df = df[mask]
df.shape

(16843, 17)

Now we get a dataframe with 16843 values, however for values like 'Otros' there are value that we dont need. Let's see them

In [7]:
df_others = df.loc[df['property_type'] == 'Otros']
print(df_others.shape)
df_others.head()

(39, 17)


title    price_S  \
id                                                                   
3367               ¡En Venta Exclusiva Casa en Camacho!  S/5840000   
4395  Vendo Local 3 Pisos Mas Azotea, Urb El Pinar, ...   S/832000   
5778             Hotel en Venta con Excelente Ubicacion  S/1647000   
6711        Venta Departamento Espalda Plaza San Miguel   S/660389   
6930        Casa en Venta en La Urb. Risso, Santa Anita  S/1168000   

       price_USD                                            address  \
id                                                                    
3367  USD1600000                    CAMACHO, CamachoLa Molina, Lima   
4395   USD225000  Urb El Pinar, Calle 14, paralela cdra 9 Av Tra...   
5778   USD450000  Av. Canta Callao, San Martín De Porres, San Ma...   
6711   USD185000  Rosa Perez Liendo , Pando Etapa IvSan Miguel, ...   
6930   USD320000  Calle Francia, San Jose De MangomarcaSanta Ani...   

     property_type                                         image_urls  \
id                                                                      
3367         Otros  {https://img10.naventcdn.com/avisos/resize/111...   
4395         Otros  {https://img10.naventcdn.com/avisos/resize/111...   
5778         Otros  {https://img10.naventcdn.com/avisos/resize/111...   
6711         Otros  {https://img10.naventcdn.com/avisos/resize/111...   
6930         Otros  {https://img10.naventcdn.com/avisos/resize/111...   

                                            description area_total area_roof  \
id                                                                             
3367  ¡Bienvenido a la exclusiva casa en venta en Ca...     1138m²    1138m²   
4395  Vendo amplio local: 3 pisos mas azotea, interc...      178m²     504m²   
5778  CONTACTAR CON ELVIS LEON AL 9.7.1.1.5.0.2.3.9....      211m²    1162m²   
6711  Dpto. 140mt. Segundo piso <br>3 dormitorios <b...      140m²     140m²   
6930  Doors te presenta esta oportunidad de negocio,...      152m²     465m²   

     bath   bed  parking antiquity  \
id                                   
3367    2   4.0      5.0        20   
4395    4   NaN      NaN        33   
5778   28  30.0      7.0        30   
6711    2   3.0      NaN        25   
6930    8  14.0      8.0        15   

                                       general_features  \
id                                                        
3367  {Jardín(es),"Centros Comerciales Cercanos","Ce...   
4395  {"Centros Comerciales Cercanos","Licencia de f...   
5778                                                NaN   
6711  {"Caseta de guardia","Centros Comerciales Cerc...   
6930  {"Baño de servicio",Terraza,"Niveles construid...   

                                               services  \
id                                                        
3367                 {"Sistema de alarma de seguridad"}   
4395  {"Uso Profesional","Uso Comercial","Sala de Re...   
5778                                                NaN   
6711  {"Guardianía/Seguridad privada","Area de lavan...   
6930  {"Uso Profesional","Guardianía/Seguridad priva...   

                                               ambience  \
id                                                        
3367                                                NaN   
4395                                         {Oficinas}   
5778                                                NaN   
6711  {Desagüe,"Sala de estar",Bodega(s),"Ingreso in...   
6930  {Patio,"Baño Propio",Bodega(s),"Ingreso indepe...   

                                           property_url  
id                                                       
3367  https://urbania.pe/inmueble/clasificado/veclot...  
4395  https://urbania.pe/inmueble/clasificado/veclot...  
5778  https://urbania.pe/inmueble/clasificado/veclot...  
6711  https://urbania.pe/inmueble/clasificado/veclot...  
6930  https://urbania.pe/inmueble/clasificado/veclot...

As we can appreciate there are values which tittle says 'Vendo Local' or 'Hotel en Venta' we dont need that values. Let's drop them out

In [8]:
pattern = '|'.join(housing_values)
pattern

'Departamento|Casa|Casa de playa|Casa de campo|Habitación|Otros'

In [9]:
id_others = df_others.loc[~df_others['title'].str.contains(pattern, case=False, regex=True)].index.to_list()
df = df[~df.index.isin(id_others)]
print(df.shape)
df.tail(2)

(16815, 17)


title   price_S  price_USD  \
id                                                                         
22414                 Venta Dpto 1er Piso y Cochera  S/383250  USD105000   
22415  Moderno Flat de 2 Dorm. en Magdalena del Mar  S/703000  USD190000   

                                       address property_type  \
id                                                             
22414  Calle Monsefú 189, San MiguelLima, Lima  Departamento   
22415      Juan de Aliaga, MagdalenaLima, Lima  Departamento   

                                              image_urls  \
id                                                         
22414  {https://img10.naventcdn.com/avisos/resize/111...   
22415  {https://img10.naventcdn.com/avisos/resize/111...   

                                             description area_total area_roof  \
id                                                                              
22414  ÁREA: 70 M2 <br><br>Ubicado en calle Monsefú c...       70m²      70m²   
22415  **Ideal para inversión** <br><br>Precioso flat...       87m²      90m²   

      bath  bed  parking antiquity  \
id                                   
22414    2  2.0      NaN         8   
22415    2  2.0      NaN         4   

                                        general_features  \
id                                                         
22414  {"Sistema contra incendios","Reposteros en coc...   
22415  {"Sistema contra incendios","Reposteros en coc...   

                                                services  \
id                                                         
22414  {"Video vigilancia","Area de lavandería","Sist...   
22415  {"Video vigilancia","Areas verdes","Línea blan...   

                                                ambience  \
id                                                         
22414          {Desagüe,"Dormitorio principal con baño"}   
22415  {Desagüe,"Sala de estar","Hall de ingreso","Ar...   

                                            property_url  
id                                                        
22414  https://urbania.pe:443/inmueble/clasificado/ve...  
22415  https://urbania.pe:443/inmueble/clasificado/ve...

Now our dataframe has only values related to housing types. Let's update that values which are like 'Otros'

In [10]:
mask = df['property_type'] == 'Otros'
other_values = df.loc[mask, 'title'].str.extract('({})'.format(pattern))[0]
df.loc[mask, 'property_type'] = other_values
print(df.shape)
df.loc[df['property_type'] == 'Otros']

(16815, 17)


Empty DataFrame
Columns: [title, price_S, price_USD, address, property_type, image_urls, description, area_total, area_roof, bath, bed, parking, antiquity, general_features, services, ambience, property_url]
Index: []

In [11]:
df.head(2)

title   price_S  price_USD  \
id                                                                           
0   Proyecto inmobilario en venta en Calle Coronel...  S/445388  USD120375   
1   Proyecto inmobilario en venta en Calle Coronel...  S/441323  USD119277   

                                              address property_type  \
id                                                                    
0   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   
1   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   

                                           image_urls  \
id                                                      
0   {https://img10.naventcdn.com/avisos/resize/111...   
1   {https://img10.naventcdn.com/avisos/resize/111...   

                                          description area_total area_roof  \
id                                                                           
0   Mozz se deriva de la palabra en italiano mozio...        NaN      45m²   
1   Mozz se deriva de la palabra en italiano mozio...       45m²      45m²   

   bath  bed  parking   antiquity  \
id                                  
0   NaN  NaN      NaN  A estrenar   
1   NaN  NaN      NaN  A estrenar   

                                     general_features          services  \
id                                                                        
0   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
1   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   

   ambience                                       property_url  
id                                                              
0       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
1       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...

Now we have the type of estates that we need. So the next step is drop that instances than not have the mandatory values, as price, address, area, number of bedroom and bathroom.

First, Let's see how many NaN value per column there are

In [12]:
df.isna().sum()

title                   0
price_S                34
price_USD            2432
address                 0
property_type           1
image_urls              0
description             1
area_total            449
area_roof            1478
bath                 2719
bed                  1832
parking             11551
antiquity               0
general_features     1429
services             5239
ambience            10095
property_url            0
dtype: int64

## Price Cleansing

There are two type of currency, Soles (Peru's local currency) and USD.

In [21]:
df['price_S'] = df['price_S'].str.extract('(\d+)').astype(float)
df['price_USD'] = df['price_USD'].str.extract('(\d+)').astype(float)
df.head(2)


title   price_S  price_USD  \
id                                                                           
0   Proyecto inmobilario en venta en Calle Coronel...  445388.0   120375.0   
1   Proyecto inmobilario en venta en Calle Coronel...  441323.0   119277.0   

                                              address property_type  \
id                                                                    
0   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   
1   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   

                                           image_urls  \
id                                                      
0   {https://img10.naventcdn.com/avisos/resize/111...   
1   {https://img10.naventcdn.com/avisos/resize/111...   

                                          description area_total area_roof  \
id                                                                           
0   Mozz se deriva de la palabra en italiano mozio...        NaN      45m²   
1   Mozz se deriva de la palabra en italiano mozio...       45m²      45m²   

   bath  bed  parking   antiquity  \
id                                  
0   NaN  NaN      NaN  A estrenar   
1   NaN  NaN      NaN  A estrenar   

                                     general_features          services  \
id                                                                        
0   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
1   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   

   ambience                                       property_url  
id                                                              
0       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  
1       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...

In [22]:
df[['price_S', 'price_USD']].isna().sum()

price_S        35
price_USD    2432
dtype: int64

Then price_S has less NaN values than price_USD. Let's drop that instance that both prices are NaN

In [23]:
mask = (df['price_S'].isna()) & (df['price_USD'].isna())
df = df[~mask]
df.shape

(16781, 17)

There prices are 0 instead of being NaN values. Let's drop them out as well. This happends because the website return a status code 404

In [27]:
mask = (df['price_S'] == 0) & (df['price_USD'].isna())
df = df[~mask]
df.shape

(16770, 17)

Now let's create a new column that contains the price only in USD.

In [28]:
df['price'] = df['price_USD']
df.head(2)

title   price_S  price_USD  \
id                                                                           
0   Proyecto inmobilario en venta en Calle Coronel...  445388.0   120375.0   
1   Proyecto inmobilario en venta en Calle Coronel...  441323.0   119277.0   

                                              address property_type  \
id                                                                    
0   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   
1   Calle Coronel Odriozola 128, San IsidroLima, Lima  Departamento   

                                           image_urls  \
id                                                      
0   {https://img10.naventcdn.com/avisos/resize/111...   
1   {https://img10.naventcdn.com/avisos/resize/111...   

                                          description area_total area_roof  \
id                                                                           
0   Mozz se deriva de la palabra en italiano mozio...        NaN      45m²   
1   Mozz se deriva de la palabra en italiano mozio...       45m²      45m²   

   bath  bed  parking   antiquity  \
id                                  
0   NaN  NaN      NaN  A estrenar   
1   NaN  NaN      NaN  A estrenar   

                                     general_features          services  \
id                                                                        
0   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   
1   {"Promotion Description","Promotion Title","Nú...  {"Areas verdes"}   

   ambience                                       property_url     price  
id                                                                        
0       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  120375.0  
1       NaN  https://urbania.pe/inmueble/proyecto/ememvein-...  119277.0

In [29]:
mask = (df['price'].isna()) | (df['price'] == 0)
df.loc[mask, 'price'] = round(df.loc[mask, 'price_S']/3.7, 1)

In [30]:
df[['price_S', 'price_USD', 'price']].isna().sum()

price_S         1
price_USD    2387
price           0
dtype: int64

In [31]:
df.shape

(16770, 18)

Then all instance has a price which currency is USD.

## Address, Area, Bed and Bath Cleansing

Into to mandatory variable which cannot be nulls, we can find the address, area, number of bedrooms and bathromms. Let's drop that instance which info is no enough


In [32]:
df[['address', 'area_total', 'area_roof', 'bath', 'bed']].isna().sum()

address          0
area_total     448
area_roof     1478
bath          2712
bed           1825
dtype: int64

This dataframe has no empty values for address field. Let's work out the others

### Areas

Let's see the NaN values for both area and if can we extract that info from the description. For instance

In [70]:
mask = (df['area_total'].isna())
df.loc[mask, 'description'].str.extract('(\d+.{1}\d*\s*(m²|m2|M\w*))').dropna()[0]

id
0             45m2
209         105 m2
289       400.86m2
293           64m2
342          147m2
           ...    
22186       264 M2
22267       653 m2
22334    100.57 m2
22337        230m2
22356         60m2
Name: 0, Length: 209, dtype: object

That means we can extract area from the description.

In [76]:
df['area'] = df['area_total']
mask = (df['area'].isna())
ids_area = df.loc[mask, 'description'].str.extract('(\d+.{1}\d*\s*(m²|m2|M\w*))').dropna().index.to_list()
df.loc[ids_area, 'area'] = df.loc[mask, 'description'].str.extract('(\d+.{1}\d*\s*(m²|m2|M\w*))').dropna()[0]
df[['area_total', 'area_roof', 'area']].isna().sum()

area_total     448
area_roof     1478
area           239
dtype: int64

Moreover, it can be if 'area_total' is NaN value, 'area_roof' could not be. In that case we define the first one as the second one.

In [84]:
mask = (df['area_total'].isna()) & ~(df['area_roof'].isna())
df.loc[mask, 'area'] = df.loc[mask,'area_roof']
df[['area_total', 'area_roof', 'area']].isna().sum()

area_total     448
area_roof     1478
area            10
dtype: int64

The 10 missing values can be dropped out because there's not other way where extract them from

In [88]:
mask = (df['area'].isna())
df = df[~mask]
df.shape

(16760, 19)

### Bed & Bath

In [95]:
df['bed'] = df['bed'].astype(float)
df[['bed']].isna().sum()

bed    1821
dtype: int64

In [143]:
mask = (df['bed'].isna())
df.loc[mask, 'description'].str.extractall('(\d*\s*[Hh]abita)')

0
id    match           
0     0         habita
1     0         habita
69    0         habita
355   0         habita
610   0         habita
...                ...
22060 2      01 Habita
22165 0         habita
22323 0         habita
      1         habita
22381 0       2 habita

[424 rows x 1 columns]

In [148]:
mask = (df['bath'].isna())
df.loc[mask, 'description'].str.extractall('(\d*\s*[Bb]año)').reset_index()[['id', 0]].groupby(['id']).sum().dropna()

0
id                     
10       baño baño baño
28               1 baño
42       baño baño baño
59       baño baño baño
69                 baño
...                 ...
22324            1 Baño
22327            2 baño
22329            1 baño
22354              baño
22368  1 baño baño baño

[1587 rows x 1 columns]

In [149]:
df[mask].shape

(2708, 19)

In [127]:
df.loc[mask, 'description'].shape

(1821,)